File: uikit/src/main/java/com/sendbird/uikit/tasks/JobTask.java
package com.sendbird.uikit.tasks;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sendbird.uikit.log.Logger;

import java.util.concurrent.Callable;

public abstract class JobTask<T> {
    @Nullable
    protected abstract T call() throws Exception;

    @NonNull
    final Callable<T> getCallable() {
        return () -> {
            try {
                return JobTask.this.call();
            } catch (Exception e) {
                Logger.e(e);
            }
            return null;
        };
    }
}

File: uikit/src/main/java/com/sendbird/uikit/internal/tasks/JobTask.kt
package com.sendbird.uikit.internal.tasks

import com.sendbird.uikit.log.Logger
import java.util.concurrent.Callable

internal abstract class JobTask<T> {
    @Throws(Exception::class)
    protected abstract fun call(): T?
    val callable: Callable<T>
        get() = Callable {
            var result: T? = null
            try {
                result = call()
            } catch (e: Exception) {
                Logger.e(e)
            }
            result
        }
}
